### ISSUE TITLE
Changing query headers returns the same cached items.

### ISSUE THREAD

**Issue by Pixelatex** [apollographql/apollo-client#8764]
**Intended outcome:**
After a query has been done, using useQuery will fetch the data from the cache each time. This is fine. However, when I change a header in the queries, I expect the cache to notice the difference and treat this as a separate query and fetch the query instead of using the cache.

**Actual outcome:**
Changing a header doesn't matter for apollo and it thinks nothing has changed in the query request.

**How to reproduce the issue:**
Set a  header via setContext and change it afterwards, the cache doesn't treat this is any difference and will serve you the same cache content again.

**Versions**

  System:
    OS: macOS 11.4
  Binaries:
    Node: 15.14.0 - ~/.nvm/versions/node/v15.14.0/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 7.7.6 - ~/.nvm/versions/node/v15.14.0/bin/npm
  Browsers:
    Chrome: 93.0.4577.63
    Firefox: 91.0.2
    Safari: 14.1.1
  npmPackages:
    @apollo/client: ^3.1.3 => 3.1.5 
    apollo-boost: ^0.4.9 => 0.4.9 
    apollo-client: ^2.6.10 => 2.6.10 
    apollo-link: ^1.2.14 => 1.2.14 
    apollo-link-http: ^1.5.17 => 1.5.17 
    apollo-link-persisted-queries: ^0.2.2 => 0.2.2 
    apollo-storybook-react: ^0.2.4 => 0.2.4

**Comment by benjamn** [comment]
Though I agree changing headers should trigger a network request, this issue seems closely related to #7650. Let me know if you disagree and I'll reopen!

**Comment by Pixelatex** [comment]
@benjamn I am re-opening this kind of as I have bumped my head on this issue again. Back then, it was decided to disable persisted queries for now and recently, we re-enabled it to get better response times and so we are once again facing this same issue.

Now, I'll try to explain my usecase here and why I believe that headers should be used in the cache identifier and considered as a unique request.

All data requires an organisation id to know for what customer you are requesting.. say.. the latest blog posts.
We could add this into every single request and pass it along but that's a lot of extra code. We instead opted to pass this organisation as a header so ALL graphql requests coming from the site used by organisation A will have their id included. This is plucked from the context in the resolver and used to get the specific data for that organisation.

This is where persisted queries is failing us. It doesn't consider a call to get blog posts from organisation A or organisation B a different request, so everyone else is getting the same posts back as the first organisation who called when the cache expired.

Long story short. Please please add the headers for a request into the unique id used to store cached responses in the cache, this way we can make use of headers and persisted queries.
